Creating plugins using iina-cli

lhc70000 opened the pull request:
**Description:**

- Added plugin-related commands to `iina-cli`.
  - `create`: See https://github.com/iina/iina-plugin-template for more description.
  - `pack`: Create a zip file with the plugin folder's content.
  - `link`: Create a symlink in IINA's plugin folder.
  - `unlink`: Remove the above symlink.
- Fixed a crash when installing local `.iinaplgz` plugins.

Recommended steps for testing:

- Make sure that `iina-cli` can create plugins using various options.
- If building is required for the generated plugin (`iina-cli` will print instructions), make sure `npm install` and `npm run build` finish successfully.
- Make sure that `link` and `unlink` work for the generated plugin folder, i.e. IINA can load the folder as a plugin correctly.
- Make sure IINA can load the zip archive generated by `pack`.

low-batt commented on iina-cli/plugin.swift:
```diff
@@ -0,0 +1,387 @@
+    }
+  }
+  
+  let inputDir = pluginDir.lastPathComponent
```
`inputDir` is not used. Xcode is warning about this.

low-batt commented on iina-cli/plugin.swift:
```diff
@@ -0,0 +1,387 @@
+  
+  var pluginDir = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
+  if #available(macOS 13.0, *) {
+    pluginDir.append(component: name)
```
I'm wondering if `available` in Xcode 13 does not properly support Ventura. The Xcode 13 build failed reporting:
```text
Users/runner/work/iina/iina/iina-cli/plugin.swift:61:15: error: value of type 'URL' has no member 'append'
    pluginDir.append(component: name)
    ~~~~~~~~~ ^~~~~~
```

saagarjha replied:
This also needs `#if MACOS_13_AVAILABLE`

low-batt commented on iina-cli/plugin.swift:
```diff
@@ -0,0 +1,387 @@
+  let templateURL = userTemplateURL ?? defaultTemplateURL
+  
+  // Use shell commands here for simplicity
+  let tmpURL = FileManager.default.temporaryDirectory
```
`temporaryDirectory` missing from macOS 10.11. The Xcode 13 build failed reporting:
```text
/Users/runner/work/iina/iina/iina-cli/plugin.swift:118:36: error: 'temporaryDirectory' is only available in macOS 10.12 or newer
  let tmpURL = FileManager.default.temporaryDirectory
                                   ^
/Users/runner/work/iina/iina/iina-cli/plugin.swift:118:36: note: add 'if #available' version check
  let tmpURL = FileManager.default.temporaryDirectory
                                   ^
```

low-batt commented on iina-cli/main.swift:
```diff
@@ -8,6 +8,51 @@
+    """
+    Usage:
+        iina-cli [arguments] [files] [-- mpv_option [...]]
+        iina-cli plugin <command> <args>
```
This command line structure seems odd to me. I'm thinking this feature does not belong in `iina-cli`. What % of current `iina-cli` users would use this feature? How often would they use the plugin commands vs. playing media? I'm thinking a separate tool for plugins would be a better solution.

saagarjha replied:
+1

low-batt reviewed (changes requested):
Changes are required if we are going to continue to support 10.11 and 10.12. See comments inline about the Xcode 13 build failures.

My first attempt to use `iina-cli` failed because I thought `create` did what `pack` does. Trouble is `iina-cli` did not tell me anything about what I did wrong:

<details>
<summary>First attempt:</summary>

```text
low-batt@gag MacOS$ ./iina-cli plugin create plugin-online-media
Usage:
    iina-cli [arguments] [files] [-- mpv_option [...]]
    iina-cli plugin <command> <args>

Arguments:
    --mpv-*:
        All mpv options are supported here, except those starting with "--no-".
        Example: --mpv-volume=20 --mpv-resume-playback=no
    --separate-windows | -w:
        Open all files in separate windows.
    --stdin, --no-stdin:
        You may also pipe to stdin directly. Sometimes iina-cli can detect
        whether stdin has file, but sometimes not. Therefore it's recommended
        to always supply --stdin when piping to iina, and --no-stdin when you
        are not intend to use stdin.
    --keep-running:
        Normally iina-cli launches IINA and quits immediately. Supply this
        option if you would like to keep it running until the main application
        exits.
    --pip:
        Enter Picture-in-Picture after opening the media.
    --help | -h:
        Print this message.

mpv Option:
    Raw mpv options without --mpv- prefix. All mpv options are supported here.
    Example:
        iina-cli test.mp4 -- --volume=20 --no-resume-playback

Plugin Commands:
    create <name>
        Create a new IINA plugin at the current directory with specified name.
    pack <dir>
        Compress a plugin folder into an .iinaplgz file.
    link <path>
        Create a symlink to the plugin folder at <path> so IINA can load it as
        a development package.
    unlink <path>
        Remove the plugin symlink from IINA's plugin folder.
low-batt@gag MacOS$ 
```
</details>

I used `pack` and was able to create the `.iinaplgz` file. I could not figure out what the `create` command does.

lhc70000 committed:
Add iina-plugin command line tool

lhc70000 commented:
@low-batt Updated. Sorry for the confusion; the `create` command was later renamed to `new` but I forgot to change the help message. Now `iina-plugin` is a standalone command.

This is the whole workflow I have tested on my machine:

```sh
# create a new plugin named "foo"
iina-cli new foo

# ...follow the prompts

# if Vue or React is required, `iina-plugin` will suggest you run` npm install` as follows
cd foo && npm install
npm build

cd -
iina-plugin link foo

# ...open iina and test if the plugin is loaded

iina-plugin unlink foo
rm -rf foo
```

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+//
+//  plugin.swift
+//  iina-cli
```
iina-plugin

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+  print(
+    """
+    Usage:
+        iina-plugin <command> <args>
```
It is desirable to follow Posix conventions which can be found [here](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01). But definitely this should match the conventions of `iina-cli`:
```bash
low-batt@gag MacOS$ ./iina-cli --help
Usage: iina-cli [arguments] [files] [-- mpv_option [...]]
```
Therefore:
- No newline after "Usage:"
- "[arguments]" instead of "\<args\>"

So maybe?:
```text
Usage: iina-plugin <command> [arguments]
```

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+
+    Plugin Commands:
+        new <name>
+            Create a new IINA plugin at the current directory with specified name.
```
Replace "at the" with "in the"

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+  var userTemplateURL: String?
+  var idxToBeDropped: [Int] = []
+  for (idx, arg) in args.enumerated() {
+    if arg.hasPrefix("--url=") {
```
The `--url` option is not documented in the help.

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+  
+  // Download the template. We uploaded the file
+  // https://github.com/iina/iina-plugin-template/archive/refs/heads/master.zip
+  // to iina.io to avoid
```
avoid what?

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+      print("  \(i + 1): \(choice)")
+    }
+    print("Input a number (default: 1): ", terminator: "")
+    if let answer = readLine() {
```
if answer.isEmpty {
        return choices[0]
      }

low-batt replied:
Because that check is missing defaulting is not working:
```text
Use a frontend framework to build the user interface?
  1: None
  2: React
  3: Vue
Input a number (default: 1): 
  1: None
  2: React
  3: Vue
Input a number (default: 1): 
```

low-batt commented on iina-plugin/main.swift:
```diff
@@ -0,0 +1,430 @@
+  print("Please run the following commands to install npm packages:")
+  print("\n  cd \(name) && npm install\n")
+  print("Use the following command to build the project:")
+  print("\n  npm run build\n")
```
`package.json` seems to be missing:
```
low-batt@gag iina-plugin-test$ /Users/low-batt/Library/Developer/Xcode/DerivedData/iina-caydzgleiqupoxbtitkkfadxksfs/Build/Products/Debug/IINA.app/Contents/MacOS/iina-plugin new testit

Include a global entry? (y/n, default: n) 
Create template for video overlay? (y/n, default: n) 
Create template for a side bar view? (y/n, default: n) y
Create template for a standalone window? (y/n, default: n) y
Use a frontend framework to build the user interface?
  1: None
  2: React
  3: Vue
Input a number (default: 1): 
  1: None
  2: React
  3: Vue
Input a number (default: 1): 
  1: None
  2: React
  3: Vue
Input a number (default: 1): 1
Use a bundler (Parcel)? (y/n, default: n) 

----------

Downloading to /var/folders/bh/sprgrpyn1mz_0f7m9_1bqzg40000gq/T/iina-cli
...Copied ./ui/sidebar/index.html
...Copied ./ui/sidebar/index.js
...Copied ./ui/window/index.html
...Copied ./ui/window/index.js
...Copied ./jsconfig.json
...Copied ./.gitignore
...Copied ./Info.json
...Copied ./src/index.js
Plugin directory created.

----------

Please run the following commands to install npm packages:

  cd testit && npm install

Use the following command to build the project:

  npm run build

Please read the plugin documentation and inspect the generated package.json for more info.
low-batt@gag iina-plugin-test$ pushd testit/
~/Documents/builds/iina-plugin-test/testit ~/Documents/builds/iina-plugin-test
low-batt@gag testit$ npm install
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /Users/low-batt/Documents/builds/iina-plugin-test/testit/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/Users/low-batt/Documents/builds/iina-plugin-test/testit/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/low-batt/.npm/_logs/2023-07-07T02_17_42_563Z-debug-0.log
low-batt@gag testit$ ls
Info.json		jsconfig.json		package-lock.json	src			ui
low-batt@gag testit$ npm run build
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /Users/low-batt/Documents/builds/iina-plugin-test/testit/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/Users/low-batt/Documents/builds/iina-plugin-test/testit/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/low-batt/.npm/_logs/2023-07-07T02_18_59_119Z-debug-0.log
low-batt@gag testit$ 
low-batt@gag testit$ ls -l
total 24
-rw-r--r--@ 1 low-batt  staff  453 Jul  6 22:17 Info.json
-rw-r--r--@ 1 low-batt  staff  132 Jul  6 22:17 jsconfig.json
-rw-r--r--  1 low-batt  staff   85 Jul  6 22:17 package-lock.json
drwxr-xr-x  3 low-batt  staff   96 Jul  6 22:17 src
drwxr-xr-x  4 low-batt  staff  128 Jul  6 22:17 ui
low-batt@gag testit$ 
```

low-batt committed:
Complete adding iina-plugin to Xcode project

This commit will:
- Change the deployment target for iina-plugin to macOS 10.13
- Add a iina-plugin.xcconfig build configuration file
- Add iina-plugin to the copy executables phase of iina

lhc70000 committed:
Remove npm instruction when not using bundler

lhc70000 committed:
Address review comments

lhc70000 commented:
Updated.

lhc70000 committed:
Add more checks when packing the package

Output the plugin at the current directory

low-batt commented:
Caught me still experimenting with this PR. I didn't find anything important. Two minor comments.

If this:
```text
Please read the plugin documentation at https://docs.iina.io.
```
Was changed to this:
```text
Please read the plugin documentation at: https://docs.iina.io
```
It would be easier for users to select just the URL as they don't have to carefully avoid the period at the end.

Possibly this should be changed to use `iina-plugin` for the directory name:
```text
Downloading to /var/folders/bh/sprgrpyn1mz_0f7m9_1bqzg40000gq/T/iina-cli
```
